builtin/logical/pki/chain_util.go

```diff
@@ -0,0 +1,1305 @@
+package pki
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+	"sort"
+
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+func prettyIssuer(issuerIdEntryMap map[issuerID]*issuerEntry, issuer issuerID) string {
+	if entry, ok := issuerIdEntryMap[issuer]; ok && len(entry.Name) > 0 {
+		return "[id:" + string(issuer) + "/name:" + entry.Name + "]"
+	}
+
+	return "[" + string(issuer) + "]"
+}
+
+func rebuildIssuersChains(ctx context.Context, s logical.Storage, referenceCert *issuerEntry /* optional */) error {
```

Comment:
There are a couple of giant functions in this file that I would love to see broken down to make reasoning about them easier, if possible. (Granted, a lot of these are detailed comments.)

Also, is it possible to use some standard-ish libraries, like [gonum](https://github.com/gonum/gonum), to do some of these graph computations?

Reply:
I'd love to see benchmark profiles for this code.

Reply:
Just as an aside, like the idea of benchmarks, but this function gets called one of two places:

 - Importing/generating an issuer,
 - Deleting an issuer.
 
And is otherwise static.

Deleting might be a little surprising for a perf hit but importing/generating I don't think is that surprising.

@sgmiller What are you thinking of for benchmarks? How many issuers until it stalls? Complexity of chains? No changes and force-trigger rebuilds? &c.

Reply:
See #15315 :-) Looks like those numbers are fairly representative, so delete overhead isn't all that much.